File: apply_translations.js
const fs = require('fs');
const path = require('path');

const wordsPath = path.join(__dirname, 'words.json');
const words = JSON.parse(fs.readFileSync(wordsPath, 'utf-8'));

// Load MyMemory results (first 200)
const myMemoryResults = JSON.parse(fs.readFileSync(path.join(__dirname, 'translation_results.json'), 'utf-8'));

// Load batch results
const batchFiles = [];
for (let i = 0; i < 8; i++) {
  const p = path.join(__dirname, `batch_${i}_done.json`);
  if (fs.existsSync(p)) {
    const batch = JSON.parse(fs.readFileSync(p, 'utf-8'));
    batchFiles.push(...batch);
    console.log(`batch_${i}_done.json: ${batch.length} items`);
  } else {
    console.log(`batch_${i}_done.json: NOT FOUND`);
  }
}

console.log(`MyMemory results: ${myMemoryResults.length}`);
console.log(`Batch results: ${batchFiles.length}`);

// Build translation map
const translations = new Map();

// Add MyMemory results (skip warnings)
for (const r of myMemoryResults) {
  if (r.status === 'ok' && r.translated && !r.translated.includes('MYMEMORY WARNING')) {
    translations.set(r.word, { english: r.english, chinese: r.translated });
  }
}

// Add batch results (override MyMemory if exists)
for (const b of batchFiles) {
  if (b.chinese && b.chinese.length > 1) {
    translations.set(b.word, { english: b.english, chinese: b.chinese });
  }
}

console.log(`Total translations available: ${translations.size}`);

// Apply to words.json
let updated = 0;
let unchanged = 0;
let alreadyGood = 0;

for (let i = 0; i < words.length; i++) {
  const w = words[i];
  
  // Skip if already has good Chinese
  if (w.examples && w.examples[0] && /[\u4e00-\u9fff]/.test(w.examples[0]) && !/@|KLAR|[A-Z]{8,}/.test(w.examples[0])) {
    alreadyGood++;
    continue;
  }
  
  // Find translation
  const trans = translations.get(w.word);
  if (trans) {
    const newExample = `${trans.english} ${trans.chinese}`;
    w.examples = [newExample];
    updated++;
  } else {
    unchanged++;
  }
}

// Backup original
fs.copyFileSync(wordsPath, path.join(__dirname, 'words.json.bak.before_translate'));

// Save
fs.writeFileSync(wordsPath, JSON.stringify(words, null, 2), 'utf-8');

console.log(`\nResults:`);
console.log(`  Already had good Chinese: ${alreadyGood}`);
console.log(`  Updated with translation: ${updated}`);
console.log(`  Unchanged (no translation): ${unchanged}`);
console.log(`  Total words: ${words.length}`);
console.log(`\nBackup saved to words.json.bak.before_translate`);

// Verify
let withCn = 0, withoutCn = 0;
for (const w of words) {
  if (w.examples && w.examples[0] && /[\u4e00-\u9fff]/.test(w.examples[0])) {
    withCn++;
  } else {
    withoutCn++;
  }
}
console.log(`\nVerification: ${withCn} with Chinese, ${withoutCn} without Chinese`);

File: apply_translations.py
import json
import sys

# Import translations
sys.path.insert(0, r'E:\Tina\自研背单词软件')
from translations_data import translations

# Load words.json
with open(r'E:\Tina\自研背单词软件\words.json', 'r', encoding='utf-8') as f:
    words_data = json.load(f)

print("Loaded " + str(len(words_data)) + " words")

# Build index: number -> word object
word_index = {}
for w in words_data:
    word_index[w['number']] = w

# Apply translations
fixed_count = 0
skipped_count = 0
errors = []

for idx, num, exIdx, en_text, zh_text in translations:
    if num not in word_index:
        errors.append("Word number " + str(num) + " not found")
        continue
    
    word_obj = word_index[num]
    examples = word_obj.get('examples', [])
    
    if exIdx >= len(examples):
        errors.append("exIdx " + str(exIdx) + " out of range for word " + str(num) + " '" + word_obj.get('word', '') + "' (only " + str(len(examples)) + " examples)")
        continue
    
    # Build new example: cleaned English + Chinese
    new_example = en_text + ' ' + zh_text
    examples[exIdx] = new_example
    fixed_count += 1

# Save
with open(r'E:\Tina\自研背单词软件\words.json', 'w', encoding='utf-8') as f:
    json.dump(words_data, f, ensure_ascii=False, indent=2)

print("Saved words.json")
print("Fixed: " + str(fixed_count) + " examples")
if errors:
    print("Errors (" + str(len(errors)) + "):")
    for e in errors:
        print("  " + e)

# Verify
with open(r'E:\Tina\自研背单词软件\words.json', 'r', encoding='utf-8') as f:
    verify = json.load(f)
print("Word count verified: " + str(len(verify)))
